File: crates/zdc-bench/js/benchmark.js
// The js-framework-benchmark workload, written five ways against one DOM.
//
// The workload is the standard one: create 1,000 rows, replace them, update
// every 10th, select a row, swap two, remove one, clear, then create 10,000
// and append 1,000 more. Every arm below performs the SAME transitions over
// the SAME row objects and must end each step with the same DOM — the
// driver compares a digest across arms and fails if they diverge, so an arm
// cannot look fast by doing less.
//
// The arms:
//
//   zd-positional   what `zdc build` emits: one cloned template per row,
//                   bindings at the holes, reconciled by `eachInto` with
//                   the interim positional key function (spec §16.6).
//   zd-identity     the same emission with the key function `record …
//                   unique` will supply. One argument differs.
//   direct          what a naive code generator would have produced:
//                   nested `elements.js` calls, one runtime call per node
//                   (spec §16.1, the rejected design).
//   vanilla         hand-written JavaScript, node by node, one listener
//                   per row — what js-framework-benchmark's `vanillajs`
//                   entry does.
//   vanilla-tuned   hand-written and hand-tuned: one parsed template
//                   cloned per row, one delegated listener for the whole
//                   list, and direct DOM edits for each operation. It is
//                   told what changed; the reactive arms are not. This is
//                   the floor §14A.2 says we lose to.
//
// React and SolidJS are absent because CI has no network and §8 forbids a
// Node dependency, so they cannot be fetched or run. What is measured here
// is what can be measured honestly without them.

// --- the row shape --------------------------------------------------------
//
// ZDC-ROW-TEMPLATE — this is the compiler's own output for
// `crates/zdc-bench/bench/row.zd`. `tests/fidelity.rs` recompiles that file
// and fails the build if this string, the walk below, or the binding
// sequence below stops matching what the emitter produces.
const ROW_HTML = '<div class="zd-row"><span> </span><button type="button"> </button><button type="button">x</button></div>';

/** Where a row's handlers go. They are never fired; they are allocated. */
const sink = {
  select() {},
  remove() {},
};

// --- arm: ZDeceptron, as emitted -----------------------------------------
//
// A fresh `template` per arm, so each arm pays for parsing its own markup
// once rather than inheriting a parse another arm already did.

function makeZdRow() {
  const $t0 = template(ROW_HTML);
  return function zdRow(item) {
    const cls = () => item().cls;
    const rowId = () => item().id;
    const rowLabel = () => item().label;
    const $r = $t0();
    // ZDC-EMITTED-BEGIN
    const $n0 = $r.firstChild;
    const $n1 = $n0.firstChild;
    const $n2 = $n1.nextSibling;
    const $n3 = $n2.nextSibling;
    bindAttr($n0, 'class', () => 'zd-row ' + (cls)());
    bindText($n1.firstChild, rowId);
    bindText($n2.firstChild, rowLabel);
    on($n2, 'click', () => sink.select(item()));
    on($n3, 'click', () => sink.remove(item()));
    // ZDC-EMITTED-END
    return $r;
  };
}

// --- arm: direct emission -------------------------------------------------
//
// One runtime call per node, nested exactly as the view nests. This is the
// shape §16.1 rejected; it is here so the rejection is measured.

function directRow(item) {
  return Row({ class: () => item().cls }, [
    Text(() => item().id),
    Button(() => item().label, { onClick: () => sink.select(item()) }),
    Button('x', { onClick: () => sink.remove(item()) }),
  ]);
}

// --- arm: hand-written vanilla, node by node ------------------------------

function vanillaRow(row) {
  const div = document.createElement('div');
  div.setAttribute('class', 'zd-row ' + row.cls);

  const span = document.createElement('span');
  const idText = document.createTextNode(String(row.id));
  span.appendChild(idText);
  div.appendChild(span);

  const label = document.createElement('button');
  label.setAttribute('type', 'button');
  const labelText = document.createTextNode(row.label);
  label.appendChild(labelText);
  label.addEventListener('click', () => sink.select(row));
  div.appendChild(label);

  const remove = document.createElement('button');
  remove.setAttribute('type', 'button');
  remove.appendChild(document.createTextNode('x'));
  remove.addEventListener('click', () => sink.remove(row));
  div.appendChild(remove);

  return { node: div, idText, labelText };
}

// --- arm: hand-tuned vanilla ---------------------------------------------
//
// The same trick the compiler uses — parse once, clone per row — plus the
// two things a person does that a general framework cannot: one delegated
// listener for the whole list, and a direct edit per operation because the
// caller already knows what changed.

function makeTunedRow() {
  let proto = null;
  return function tunedRow(row) {
    if (proto === null) {
      const element = document.createElement('template');
      element.innerHTML = ROW_HTML;
      proto = element.content;
    }
    const fragment = proto.cloneNode(true);
    const div = fragment.firstChild;
    const idText = div.firstChild.firstChild;
    const labelText = div.firstChild.nextSibling.firstChild;
    idText.nodeValue = String(row.id);
    labelText.nodeValue = row.label;
    return { node: div, idText, labelText, fragment };
  };
}

// --- the arms -------------------------------------------------------------

function reactiveArm(renderRow, keyOf) {
  let write = null;
  return {
    mount(host) {
      const [list, set] = signal([]);
      write = set;
      const region = anchors();
      const start = region.firstChild;
      const end = region.lastChild;
      host.appendChild(region);
      eachInto(start, end, list, keyOf, renderRow);
    },
    // The reactive arms are handed the new list and nothing else. Working
    // out what changed is the framework's job, which is the whole point of
    // the comparison.
    apply(next) {
      write(next);
    },
  };
}

function vanillaArm() {
  let host = null;
  let rows = [];
  return {
    mount(h) {
      host = h;
    },
    apply(next, hint) {
      switch (hint.kind) {
        case 'create': {
          host.replaceChildren();
          rows = [];
          const fragment = document.createDocumentFragment();
          for (const row of next) {
            const record = vanillaRow(row);
            rows.push(record);
            fragment.appendChild(record.node);
          }
          host.appendChild(fragment);
          break;
        }
        case 'append': {
          const fragment = document.createDocumentFragment();
          for (const row of next.slice(next.length - hint.count)) {
            const record = vanillaRow(row);
            rows.push(record);
            fragment.appendChild(record.node);
          }
          host.appendChild(fragment);
          break;
        }
        case 'update': {
          for (const index of hint.indices) {
            rows[index].labelText.nodeValue = next[index].label;
          }
          break;
        }
        case 'select': {
          if (hint.previous >= 0) {
            rows[hint.previous].node.setAttribute('class', 'zd-row');
          }
          rows[hint.index].node.setAttribute('class', 'zd-row danger');
          break;
        }
        case 'swap': {
          const a = rows[hint.a].node;
          const b = rows[hint.b].node;
          const after = b.nextSibling;
          host.insertBefore(b, a);
          host.insertBefore(a, after);
          const record = rows[hint.a];
          rows[hint.a] = rows[hint.b];
          rows[hint.b] = record;
          break;
        }
        case 'remove': {
          rows[hint.index].node.remove();
          rows.splice(hint.index, 1);
          break;
        }
        case 'clear': {
          host.replaceChildren();
          rows = [];
          break;
        }
        default:
          throw new Error('unknown hint ' + hint.kind);
      }
    },
  };
}

function tunedArm() {
  const tunedRow = makeTunedRow();
  let host = null;
  let rows = [];
  return {
    mount(h) {
      host = h;
      // One listener for every row that will ever exist.
      host.addEventListener('click', () => {});
    },
    apply(next, hint) {
      switch (hint.kind) {
        case 'create': {
          host.replaceChildren();
          rows = [];
          const fragment = document.createDocumentFragment();
          for (const row of next) {
            const record = tunedRow(row);
            rows.push(record);
            fragment.appendChild(record.fragment);
          }
          host.appendChild(fragment);
          break;
        }
        case 'append': {
          const fragment = document.createDocumentFragment();
          for (const row of next.slice(next.length - hint.count)) {
            const record = tunedRow(row);
            rows.push(record);
            fragment.appendChild(record.fragment);
          }
          host.appendChild(fragment);
          break;
        }
        case 'update': {
          for (const index of hint.indices) {
            rows[index].labelText.nodeValue = next[index].label;
          }
          break;
        }
        case 'select': {
          if (hint.previous >= 0) {
            rows[hint.previous].node.setAttribute('class', 'zd-row');
          }
          rows[hint.index].node.setAttribute('class', 'zd-row danger');
          break;
        }
        case 'swap': {
          const a = rows[hint.a].node;
          const b = rows[hint.b].node;
          const after = b.nextSibling;
          host.insertBefore(b, a);
          host.insertBefore(a, after);
          const record = rows[hint.a];
          rows[hint.a] = rows[hint.b];
          rows[hint.b] = record;
          break;
        }
        case 'remove': {
          rows[hint.index].node.remove();
          rows.splice(hint.index, 1);
          break;
        }
        case 'clear': {
          host.replaceChildren();
          rows = [];
          break;
        }
        default:
          throw new Error('unknown hint ' + hint.kind);
      }
    },
  };
}

// --- the data -------------------------------------------------------------
//
// Built once and shared by every arm, so identity keying sees the same
// object identities the other arms see and no arm gets a different list.

let nextRowId = 1;

function build(count) {
  const out = [];
  for (let i = 0; i < count; i += 1) {
    out.push({ id: nextRowId, label: 'row ' + nextRowId, cls: '' });
    nextRowId += 1;
  }
  return out;
}

const FIRST_1K = build(1000);
const SECOND_1K = build(1000);
const TEN_K = build(10000);
const EXTRA_1K = build(1000);

function replaceAt(list, index, row) {
  const out = list.slice();
  out[index] = row;
  return out;
}

const UPDATE_INDICES = [];
for (let i = 0; i < SECOND_1K.length; i += 10) UPDATE_INDICES.push(i);

const UPDATED_1K = SECOND_1K.slice();
for (const index of UPDATE_INDICES) {
  UPDATED_1K[index] = {
    id: SECOND_1K[index].id,
    label: SECOND_1K[index].label + ' !!!',
    cls: SECOND_1K[index].cls,
  };
}

const SELECTED_1K = replaceAt(UPDATED_1K, 500, {
  id: UPDATED_1K[500].id,
  label: UPDATED_1K[500].label,
  cls: 'danger',
});

const SWAPPED_1K = SELECTED_1K.slice();
{
  const a = SWAPPED_1K[1];
  SWAPPED_1K[1] = SWAPPED_1K[998];
  SWAPPED_1K[998] = a;
}

const REMOVED_1K = SWAPPED_1K.slice();
REMOVED_1K.splice(4, 1);

const ELEVEN_K = TEN_K.concat(EXTRA_1K);

const STEPS = [
  { name: 'create 1,000 rows', next: FIRST_1K, hint: { kind: 'create' } },
  { name: 'replace 1,000 rows', next: SECOND_1K, hint: { kind: 'create' } },
  {
    name: 'update every 10th row',
    next: UPDATED_1K,
    hint: { kind: 'update', indices: UPDATE_INDICES },
  },
  {
    name: 'select a row',
    next: SELECTED_1K,
    hint: { kind: 'select', index: 500, previous: -1 },
  },
  { name: 'swap two rows', next: SWAPPED_1K, hint: { kind: 'swap', a: 1, b: 998 } },
  { name: 'remove a row', next: REMOVED_1K, hint: { kind: 'remove', index: 4 } },
  { name: 'clear 999 rows', next: [], hint: { kind: 'clear' } },
  { name: 'create 10,000 rows', next: TEN_K, hint: { kind: 'create' } },
  {
    name: 'append 1,000 to 10,000',
    next: ELEVEN_K,
    hint: { kind: 'append', count: 1000 },
  },
  { name: 'clear 11,000 rows', next: [], hint: { kind: 'clear' } },
];

// --- comparing the arms ---------------------------------------------------

/**
 * A structural digest of the rendered DOM.
 *
 * Comments are skipped: the reactive arms carry two anchor comments the
 * vanilla arms have no reason to create, and that difference is reported
 * in the node counts rather than smuggled into a correctness comparison.
 * `class` is compared as the unordered token set a browser treats it as,
 * so `"zd-row "` and `"zd-row"` are the same class list — which they are.
 */
function digest(node) {
  let hash = 0x811c9dc5;
  const push = (text) => {
    for (let i = 0; i < text.length; i += 1) {
      hash ^= text.charCodeAt(i);
      hash = Math.imul(hash, 0x01000193) >>> 0;
    }
  };
  const visit = (n) => {
    if (n.kind === 'comment') return;
    if (n.kind === 'text') {
      push('#');
      push(n.nodeValue);
      return;
    }
    if (n.kind === 'element') {
      push('<');
      push(n.tagName);
      for (const name of Object.keys(n.attributes).sort()) {
        push(name);
        push('=');
        push(
          name === 'class'
            ? n.attributes[name].split(/\s+/).filter(Boolean).sort().join(' ')
            : n.attributes[name]
        );
      }
    }
    for (const child of n.childNodes) visit(child);
    if (n.kind === 'element') push('>');
  };
  visit(node);
  return hash;
}

function elementCount(node) {
  let total = 0;
  for (const child of node.childNodes) if (child.kind === 'element') total += 1;
  return total;
}

// --- the driver -----------------------------------------------------------

const REPORT = [];

function record(arm, step, host) {
  const counts = snapshot();
  const fields = [
    'rows=' + elementCount(host),
    'digest=' + digest(host),
    'crossings=' + totalCrossings(counts.crossings),
  ];
  for (const key of COUNTERS) {
    if (counts.crossings[key] !== 0) fields.push('cross.' + key + '=' + counts.crossings[key]);
    if (counts.work[key] !== 0) fields.push('work.' + key + '=' + counts.work[key]);
  }
  for (const key of REACTIVE) {
    if (counts.reactive[key] !== 0) fields.push('reactive.' + key + '=' + counts.reactive[key]);
  }
  REPORT.push('RESULT\t' + arm + '\t' + step + '\t' + fields.join(','));
}

function runArm(name, arm) {
  const host = document.createElement('div');

  // Mounting and the first row are measured on their own: parsing the
  // template is a one-time cost paid at the first clone, and charging it
  // to `create 1,000 rows` would misattribute it.
  resetCounts();
  arm.mount(host);
  arm.apply([FIRST_1K[0]], { kind: 'create' });
  record(name, 'mount and render one row', host);

  arm.apply([], { kind: 'clear' });

  for (const step of STEPS) {
    resetCounts();
    arm.apply(step.next, step.hint);
    record(name, step.name, host);
  }
}

runArm('zd-positional', reactiveArm(makeZdRow(), byPosition));
runArm('zd-identity', reactiveArm(makeZdRow(), (item) => item.id));
runArm('direct', reactiveArm(directRow, byPosition));
runArm('vanilla', vanillaArm());
runArm('vanilla-tuned', tunedArm());

REPORT.join('\n');

File: crates/zdc-bench/tests/benchmark.rs
//! The benchmark gate.
//!
//! §14A.4: *"Regressions in these numbers are build failures, not
//! observations."* This file is where that sentence is enforced. Three kinds
//! of assertion, in order of what they protect:
//!
//! 1. **Every arm renders the same DOM.** Without this the numbers mean
//!    nothing, because an arm can always be fast by being wrong — which is
//!    exactly the failure §16.6 found in the previous reconciler, where
//!    update and swap were silent no-ops.
//! 2. **The claims in §14A and §16.1 hold**, expressed as inequalities with
//!    headroom, so a code generator that stops paying for itself fails the
//!    build rather than being noticed later.
//! 3. **The committed table still matches the code.** Any change to any
//!    number, in any direction, fails until `BENCHMARKS.md` is regenerated.
//!
//! The workload takes about two minutes and every test in this binary shares
//! one run of it.

use std::sync::OnceLock;

use zdc_bench::{bundle_sizes, generated_section, Report, END_MARKER, START_MARKER};

fn report() -> &'static Report {
    static REPORT: OnceLock<Report> = OnceLock::new();
    REPORT.get_or_init(zdc_bench::run)
}

/// The reorder measurement, shared by every test in this binary for the
/// reason the workload above is: it is a second run of the runtime and it
/// costs about as much again.
fn reorder() -> &'static Report {
    static REPORT: OnceLock<Report> = OnceLock::new();
    REPORT.get_or_init(zdc_bench::run_reorder)
}

/// The reconciler that ships, and the one it replaced.
const LIS: &str = "lis";
const CURSOR: &str = "cursor";

const ZD: &str = "zd-positional";
const IDENTITY: &str = "zd-identity";
const DIRECT: &str = "direct";
const VANILLA: &str = "vanilla";
const TUNED: &str = "vanilla-tuned";

const CREATE_10K: &str = "create 10,000 rows";
const CREATE_1K: &str = "create 1,000 rows";

/// Nothing below means anything unless the arms agree on what they rendered.
///
/// The digest is the element tree, its text, and its attributes, with
/// `class` compared as the token set a browser treats it as. An arm that
/// skipped an update, moved the wrong node, or left a stale row behind
/// fails here first.
#[test]
fn every_arm_renders_the_same_dom_at_every_step() {
    let report = report();
    let arms = report.arms();
    assert_eq!(arms.len(), 5, "expected five arms, got {arms:?}");

    for step in report.steps() {
        let first = report.find(arms[0], step);
        for arm in &arms[1..] {
            let other = report.find(arm, step);
            assert_eq!(
                other.get("digest"),
                first.get("digest"),
                "after `{step}`, `{arm}` rendered a different DOM from `{}`. \
                 One of them is wrong; the counts are meaningless until they agree.",
                arms[0]
            );
            assert_eq!(
                other.get("rows"),
                first.get("rows"),
                "after `{step}`, `{arm}` has a different row count from `{}`",
                arms[0]
            );
        }
    }
}

/// The workload really did render what it claims to have rendered.
///
/// A digest comparison passes trivially if every arm renders nothing.
#[test]
fn the_workload_rendered_the_rows_it_says_it_did() {
    let report = report();
    let arms = report.arms();
    assert_eq!(arms.len(), 5, "five arms are measured, got {arms:?}");
    for arm in arms {
        assert_eq!(report.find(arm, CREATE_1K).get("rows"), 1_000);
        assert_eq!(report.find(arm, CREATE_10K).get("rows"), 10_000);
        assert_eq!(
            report.find(arm, "append 1,000 to 10,000").get("rows"),
            11_000
        );
        assert_eq!(report.find(arm, "remove a row").get("rows"), 999);
        assert_eq!(report.find(arm, "clear 11,000 rows").get("rows"), 0);
    }
}

/// §16.1: template cloning against the direct emission it was chosen over.
///
/// The spec claims *"4.2× fewer DOM API crossings"* on this row shape.
/// Measured here it is 3.1×, which is a real advantage and not the claimed
/// one — see `BENCHMARKS.md`. The gate is set at 2× rather than at the
/// measured 3.1× because the ratio depends on how many holes and handlers a
/// row has, and a benchmark that fails when someone adds an attribute to the
/// row is a benchmark people delete. At 2× the architectural claim is still
/// falsifiable: template cloning that stopped being worth its complexity
/// could not pass it.
#[test]
fn template_cloning_halves_the_dom_crossings_of_direct_emission() {
    let report = report();
    for step in [CREATE_1K, CREATE_10K] {
        let template = report.find(ZD, step).get("crossings");
        let direct = report.find(DIRECT, step).get("crossings");
        assert!(
            template * 2 <= direct,
            "`{step}`: template cloning made {template} DOM crossings and direct emission \
             made {direct}. §16.1 chose template cloning over direct emission on this \
             number; at less than 2× the choice no longer pays for itself."
        );
    }
}

/// §16.1: *"1,000 fewer effect allocations"* at N=1,000. Measured exactly.
///
/// One effect per hole against one per node. The gate is the spec's own
/// claim, stated per row so it holds at both sizes.
#[test]
fn template_cloning_allocates_one_fewer_effect_per_row() {
    let report = report();
    for step in [CREATE_1K, CREATE_10K] {
        let rows = report.find(ZD, step).get("rows");
        let template = report.find(ZD, step).get("reactive.effect");
        let direct = report.find(DIRECT, step).get("reactive.effect");
        assert!(
            direct - template >= rows,
            "`{step}`: {template} effects against direct emission's {direct} over {rows} rows. \
             §16.1 claims at least one fewer effect per row."
        );
        assert!(
            template <= rows * 3,
            "`{step}`: {template} effects for {rows} rows. The row has three holes, so three \
             effects per row is the ceiling; more means a binding is being created that the \
             emitter does not need."
        );
    }
}

/// §14A.2: *"A hand-tuned vanilla-JS micro-app will beat us. This is the one
/// comparison we lose."*
///
/// It still does, and this test says so out loud. If it ever fails because
/// the emitter got better, the spec is what needs changing — which is the
/// point of writing the claim down as an assertion.
#[test]
fn hand_tuned_vanilla_is_still_the_floor() {
    let report = report();
    let tuned = report.find(TUNED, CREATE_10K).get("crossings");
    let template = report.find(ZD, CREATE_10K).get("crossings");
    assert!(
        tuned < template,
        "hand-tuned vanilla made {tuned} DOM crossings and the emitted code made {template}. \
         §14A.2 states plainly that hand-tuned vanilla beats us; if that has stopped being \
         true, the spec is now wrong and should be corrected rather than this test relaxed."
    );
    // 2.5×, against 1.75× measured. The ceiling is here to catch the loss
    // widening by an order of magnitude, not to fail the build over one
    // extra per-row write — the golden table already catches that exactly.
    assert!(
        template * 2 <= tuned * 5,
        "the emitted code made {template} DOM crossings against hand-tuned vanilla's {tuned}. \
         §14A.2 concedes the loss but calls it a micro-app effect that does not generalise; \
         past 2.5× at 10,000 rows it has generalised, and the concession is understated."
    );
}

/// What §14A.2 does *not* claim, and the measurement supports: emitted code
/// beats hand-written vanilla that is not hand-tuned — the node-by-node
/// style js-framework-benchmark's own `vanillajs` entry uses.
#[test]
fn emitted_code_beats_vanilla_written_node_by_node() {
    let report = report();
    let vanilla = report.find(VANILLA, CREATE_10K).get("crossings");
    let template = report.find(ZD, CREATE_10K).get("crossings");
    assert!(
        template * 2 <= vanilla,
        "emitted {template} DOM crossings against node-by-node vanilla's {vanilla}; \
         template cloning should be at least twice as frugal as building each node by hand."
    );
}

/// §16.6's cost table, which is the honest part of the interim keying
/// decision. Both halves are gated: what identity keying buys, and what
/// positional keying costs until `record … unique` lands.
#[test]
fn the_keying_costs_are_the_ones_the_spec_admits_to() {
    let report = report();

    // Identity keying: a removal is one call and nothing else moves. This is
    // what R1's two-pass retire bought (994 moves → 0).
    let removal = report.find(IDENTITY, "remove a row");
    assert!(
        removal.get("crossings") <= 2,
        "removing one row under identity keying cost {} DOM crossings; the two-pass retire \
         in §16.2 R1 makes it one `removeChild` and no moves.",
        removal.get("crossings")
    );

    // Identity keying: a two-row swap moved 997 nodes at N=1,000 until the
    // longest-increasing-subsequence reconciler §16.10 scheduled landed.
    // It is two now, and two is the minimum — the two rows that changed
    // places. Pinned exactly rather than ranged: there is no row shape or
    // list size this number depends on, so any other value is a defect
    // rather than a drift, and `reordering_moves_the_fewest_rows_it_can`
    // is what says so at three sizes and in four shapes.
    let swap = report.find(IDENTITY, "swap two rows");
    assert_eq!(
        swap.get("cross.insertBefore"),
        2,
        "a two-row swap under identity keying moved {} nodes. Exchanging two rows needs two \
         moves; more means the reconciler stopped computing a minimal move set.",
        swap.get("cross.insertBefore")
    );

    // Positional keying: identity is the slot, so a removal rewrites every
    // row after it. This is the number that makes `record … unique` urgent
    // rather than nice to have.
    let shifted = report.find(ZD, "remove a row");
    assert!(
        shifted.get("crossings") >= 1_000,
        "removing one row under positional keying cost only {} DOM crossings. That would be \
         good news and it would mean §16.6's account of positional keying is out of date.",
        shifted.get("crossings")
    );
    assert!(
        shifted.get("crossings") <= 4_000,
        "removing one row under positional keying cost {} DOM crossings, up from the 2,986 \
         measured. Positional keying is already the worst number in this suite.",
        shifted.get("crossings")
    );
}

/// Nothing below means anything unless both reconcilers agree on what they
/// rendered. A reconciler that moved fewer nodes by leaving the list in the
/// wrong order would pass every count in this file and fail here.
#[test]
fn both_reconcilers_leave_the_list_in_the_same_order() {
    let reorder = reorder();
    let steps = reorder.steps();
    assert_eq!(steps.len(), 12, "four shapes at three sizes: {steps:?}");
    for step in steps {
        let lis = reorder.find(LIS, step);
        let cursor = reorder.find(CURSOR, step);
        assert_eq!(
            lis.get("digest"),
            cursor.get("digest"),
            "after `{step}` the two reconcilers rendered different orders. \
             One of them is wrong; the move counts are meaningless until they agree."
        );
        assert_eq!(lis.get("rows"), cursor.get("rows"), "after `{step}`");
    }
}

/// §16.10, issue #207: *"Identity-keyed reordering is O(n) moves until the
/// LIS reconciler lands."* It has landed, and this is the measurement that
/// says so — the exact move set every shape costs, at every size.
///
/// Exact rather than ranged. A minimal move set is a combinatorial fact
/// about the permutation and not a property of the row shape, the engine
/// or the list's contents, so there is nothing here for headroom to
/// absorb: any other number is a defect.
#[test]
fn reordering_moves_the_fewest_rows_it_can() {
    let reorder = reorder();
    // (shape, N, minimal moves, what the cursor walk cost)
    let expected = [
        // Two rows change places, so two rows move — at every size. This
        // is the row §16.6 measured at 997 and scheduled the fix for.
        ("swap two rows", 100, 2, 97),
        ("swap two rows", 1000, 2, 997),
        ("swap two rows", 5000, 2, 4997),
        // One row is out of place, and the cursor walk already found this
        // one: it is here so the win is not overstated.
        ("move the last row to the front", 100, 1, 1),
        ("move the last row to the front", 1000, 1, 1),
        ("move the last row to the front", 5000, 1, 1),
        // A permutation is not all a real update is. Two moves for the
        // swap, one for the row appended at the end, one for the row that
        // shifted into the gap the removal left.
        ("remove one, add one, swap two", 100, 4, 98),
        ("remove one, add one, swap two", 1000, 4, 998),
        ("remove one, add one, swap two", 5000, 4, 4998),
        // The worst case, and the reason this is a minimal move set rather
        // than a small one: a reversal has no increasing subsequence longer
        // than one row, so n - 1 moves is optimal and there is nothing to
        // save. An implementation that beat this would be wrong.
        ("reverse the whole list", 100, 99, 99),
        ("reverse the whole list", 1000, 999, 999),
        ("reverse the whole list", 5000, 4999, 4999),
    ];
    for (shape, size, minimal, walked) in expected {
        let step = format!("{shape} at N={size}");
        assert_eq!(
            reorder.find(LIS, &step).get("moves"),
            minimal,
            "`{step}` should cost {minimal} moves"
        );
        assert_eq!(
            reorder.find(CURSOR, &step).get("moves"),
            walked,
            "`{step}` cost the cursor walk a different number than the {walked} recorded; \
             the before column in BENCHMARKS.md is then stale"
        );
    }
}

/// The order-of-growth claim, stated as the only thing a benchmark can
/// honestly say about one: the count stopped depending on the list.
///
/// A single size cannot distinguish O(1) from O(n) — 2 moves out of 1,000
/// and 2 moves out of 2 are the same number. Three sizes spanning 50× can:
/// the reconciler that ships costs the same at all three, and the one it
/// replaced costs fifty times more at the largest than at the smallest.
#[test]
fn the_cost_of_a_reorder_no_longer_grows_with_the_list() {
    let reorder = reorder();
    for shape in ["swap two rows", "remove one, add one, swap two"] {
        let at = |arm: &str, size: usize| {
            reorder
                .find(arm, &format!("{shape} at N={size}"))
                .get("moves")
        };
        assert_eq!(
            (at(LIS, 100), at(LIS, 1000)),
            (at(LIS, 1000), at(LIS, 5000)),
            "`{shape}` cost the LIS reconciler {}, {} and {} moves at N=100, 1,000 and 5,000. \
             The move set for this shape is the same size whatever the list's length, so a \
             count that varies with it means the reconciler is walking the list rather than \
             the moves.",
            at(LIS, 100),
            at(LIS, 1000),
            at(LIS, 5000)
        );
        assert!(
            at(CURSOR, 5000) >= at(CURSOR, 100) * 40,
            "`{shape}` cost the cursor walk {} moves at N=100 and {} at N=5,000. It is the \
             linear arm; if it has stopped being linear it is no longer the algorithm this \
             change replaced and the before column is measuring something else.",
            at(CURSOR, 100),
            at(CURSOR, 5000)
        );
    }
}

/// The one number where a general reconciler is beaten by an idiom rather
/// than by tuning: emptying a list.
///
/// A framework retires rows one at a time; `replaceChildren()` is one call
/// whatever the length. Gated so the O(n) teardown is visible rather than
/// forgotten.
#[test]
fn clearing_a_list_is_linear_for_every_reactive_arm() {
    let report = report();
    for arm in [ZD, IDENTITY, DIRECT] {
        let clear = report.find(arm, "clear 11,000 rows");
        assert_eq!(
            clear.get("cross.removeChild"),
            11_000,
            "`{arm}` cleared 11,000 rows with {} `removeChild` calls",
            clear.get("cross.removeChild")
        );
    }
    for arm in [VANILLA, TUNED] {
        assert_eq!(
            report.find(arm, "clear 11,000 rows").get("crossings"),
            1,
            "`{arm}` should clear the list with one `replaceChildren()`"
        );
    }
}

/// §14A.4 also asks for bundle size. Bytes as shipped: there is no minifier
/// in the pipeline, so these are the real numbers and not a projection.
///
/// The ceilings are round numbers roughly 50% above what is emitted today.
/// They are not a target; they exist so that a code generator that starts
/// emitting a helper per node, or a runtime that grows a framework inside
/// it, fails the build.
#[test]
fn the_emitted_bundle_and_the_runtime_stay_small() {
    for size in bundle_sizes() {
        assert!(
            size.client_js <= 2_048,
            "`{}` emitted {} bytes of client.js; the ceiling is 2,048",
            size.name,
            size.client_js
        );
    }
    let runtime = zdc_runtime::SIGNAL_JS.len() + zdc_runtime::DOM_JS.len();
    assert!(
        runtime <= 24_576,
        "the runtime a bundle links against is {runtime} bytes; the ceiling is 24,576. \
         It is unminified and heavily commented, so this is not a byte-count contest — \
         it is a check that no framework has grown inside it."
    );
}

/// The committed table is generated from the measurements, so a number in
/// the repository that disagrees with the code is a build failure.
///
/// Regenerate with `ZDC_BLESS=1 cargo test -p zdc-bench`.
#[test]
fn the_committed_results_match_the_measurements() {
    let path = zdc_bench::repository_path("BENCHMARKS.md");
    let committed = std::fs::read_to_string(&path)
        .unwrap_or_else(|e| panic!("reading {}: {e}", path.display()));
    let generated = generated_section(report(), reorder());

    let start = committed
        .find(START_MARKER)
        .unwrap_or_else(|| panic!("{} has no `{START_MARKER}`", path.display()))
        + START_MARKER.len();
    let end = committed
        .find(END_MARKER)
        .unwrap_or_else(|| panic!("{} has no `{END_MARKER}`", path.display()));
    let existing = committed[start..end].trim_matches('\n');

    if existing == generated.trim_matches('\n') {
        return;
    }

    if std::env::var_os("ZDC_BLESS").is_some() {
        let rewritten = format!(
            "{}{START_MARKER}\n\n{}\n{}",
            &committed[..start - START_MARKER.len()],
            generated.trim_matches('\n'),
            &committed[end..]
        );
        std::fs::write(&path, rewritten).expect("rewriting BENCHMARKS.md");
        panic!("BENCHMARKS.md has been regenerated. Review the diff and commit it.");
    }

    panic!(
        "BENCHMARKS.md no longer matches the measurements. §14A.4 makes that a build failure, \
         not an observation. Inspect the change, and if it is intended run \
         `ZDC_BLESS=1 cargo test -p zdc-bench` to regenerate the table.\n\n\
         committed:\n{existing}\n\nmeasured:\n{generated}"
    );
}
